### PR TITLE
added fast mode

### PR DESCRIPTION
**Pull request by wouterdb**
# Description

This PR contains the basic mechanisms for a '--fast' option on the tests.

The fast mode:

- reduces parameterized sets to a single (random) item
- skips slow tests

The idea would be

- run fast mode on commit
- run full mode nightly (or on demand)

# Todo

- [ ] roll out annotations
- [ ] update jenkinsfile

### CODE REVIEW

**Review by bartv — APPROVED** [review]
The server_multi fixture is a good candidate for this one.

File: tests/conftest.py [hunk 1]
```diff
@@ -87,6 +87,46 @@
...
+    """
+    For each test marked as parametrize_any run it either as
+    1. if not in fast mode: as if annotated with @parametrize
+    2. if in fast mode: with only one parameter combination, chosen randomly
+    """
+
+    is_fast = metafunc.config.getoption("fast")
+    for marker in metafunc.definition.iter_markers(name="parametrize_any"):
+        variations = len(marker.args[1])
+        if not is_fast or variations<=2:
```
Comment (arnaudsjs) [inline tests/conftest.py:109]: Why `<=2`?

**Review by arnaudsjs — APPROVED** [review]
Nice. I don't think I ever applied the `slowtest` label on a test case, so we might have to go the test suite and set the marker were required.

**Comment by wouterdb** [comment]
@bartv that is a parameterized fixture, no clue how I work with that (but server_multi is only used on 3 test, and I think they are all marked as slow)

File: tests/conftest.py (hunk 1 above)
Comment (wouterdb) [reply tests/conftest.py:109]: if there are 0 or 1 variations, we don't touch it?

**Comment by bartv** [comment]
> @bartv that is a parameterized fixture, no clue how I work with that (but server_multi is only used on 3 test, and I think they are all marked as slow)

ok, then we have made that optimization already

**Comment by wouterdb** [comment]
I opened a second branch for placing the actual annotation https://jenkins.inmanta.com/job/core/job/inmanta-core/job/issue%252Ffast-mode/

File: tests/conftest.py (hunk 1 above)
Comment (arnaudsjs) [reply tests/conftest.py:109]: But now it's implemented as _"if there are 0, 1 or 2 variations, we don't touch it"_. Right?

Comment (arnaudsjs) [reply tests/conftest.py:109]:
```suggestion
        if not is_fast or variations<2:
```

Comment (wouterdb) [reply tests/conftest.py]: indeed, this is wrong

**Review by sanderr — APPROVED** [review]
I find the concept of `parametrize_any` a bit scary but I guess we'll have to wait and see what it gives.

**Comment by wouterdb** [comment]
@sanderr the default is run everything, so for local testing it doesn't matter. 

I also made it an explicitly different annotation, so the idea of `parametrize_any`is to use it on tests where the parameters merely increase the variation in the test. I think the parameter should be the ratio `additional-coverage/runtime`

**Comment by arnaudsjs** [comment]
@wouterdb Should we also introduce a check that makes the tests fail when `@parametrize` is used instead of `@parametrize_any`? Currently it's easy to forget about the fact that you should be using `@parametrize_any`

**Comment by wouterdb** [comment]
@arnaudsjs I did that on purpose, I could have just changed the behavior of `@parametrize`, but I want to leave the choice
- if the test is fast and really test different things for each parameter, `@parametrize` is still the way to go
- if the parameters only slightly increase the coverage of the test (some different exception path,...) or they are fairly slow, it would advise `@parametrize_any`.
- if the test comes out as very very slow, it is `@paramterize` and `@slowtest` 

(perhaps I have to add those guidelines to the docs somewhere)



If the

**Pull request closed without merging**